lib/Pg/Reindex.pm

```diff
@@ -0,0 +1,783 @@
+                '[' . join( '], [', @not_validated ) . ']', <<'SQL';
+UPDATE reindex.log
+   SET status='failed: some constraints could not be validated: ',
+       tm_taken=now()-tstmp
+ WHERE id=$1
+SQL
+        } else {
+            query '', $log_id[1]->[0]->[0], <<'SQL';
+UPDATE reindex.log
+   SET status='done',
+       tm_taken=now()-tstmp
+ WHERE id=$1
+SQL
+        }
+    }
+
+    return 1;
+}
+
+sub init {
+
+}
+
+sub prepare {
+
+    $dbh = shift;
+    my @opt_namespaces = @{ shift @_ };
+    my @opt_tables     = @{ shift @_ };
+    my @opt_indexes    = @{ shift @_ };
+
+    my $qual;
+    my @param;
+
+    query '', 'SET client_min_messages TO WARNING';
+
+    query '', 'CREATE SCHEMA IF NOT EXISTS reindex';
+
+    query '', <<'SQL';
+CREATE TABLE IF NOT EXISTS reindex.log(
+  id BIGSERIAL PRIMARY KEY,
+  tstmp TIMESTAMP,
+  nspname NAME,
+  tblname NAME,
+  idxname NAME,
+  sz_before BIGINT,
+  sz_after BIGINT,
+  status TEXT,
+  tm_taken INTERVAL
+)
+SQL
+
+    query '', <<'SQL';
+CREATE UNLOGGED TABLE IF NOT EXISTS reindex.worklist(
+  ord SERIAL,
+  idx OID,
+  status TEXT
+)
+SQL
+
+    if (@opt_namespaces) {
+        $qual .= "   AND n.nspname IN ("
+            . join( ', ', ('?') x ( 0 + @opt_namespaces ) ) . ")\n";
+        push @param, @opt_namespaces;
+    } else {
+        $qual .= <<'SQL';
+   AND n.nspname !~ '^pg_'
+   AND n.nspname <> 'information_schema'
+   AND n.nspname <> 'reindex'
+SQL
+    }
+
+    if (@opt_tables) {
+        $qual .= "   AND EXISTS(SELECT 1
+                FROM pg_catalog.pg_class xc
+                JOIN pg_catalog.pg_index xi ON xc.oid=xi.indrelid
+               WHERE xi.indexrelid=c.oid
+                 AND xc.relname IN ("
+            . join( ', ', ('?') x ( 0 + @opt_tables ) ) . "))\n";
+        push @param, @opt_tables;
+    }
+
+    if (@opt_indexes) {
+        $qual .= "   AND c.relname IN ("
+            . join( ', ', ('?') x ( 0 + @opt_indexes ) ) . ")\n";
+        push @param, @opt_indexes;
+    }
+
+    query '', 'TRUNCATE reindex.worklist';
+    query '',
+        q{SELECT pg_catalog.setval('reindex.worklist_ord_seq', 1, false)};
+
+    query '', @param, <<'SQL'. $qual . <<'SQL';
+INSERT INTO reindex.worklist(idx, status)
+SELECT c.oid, 'planned'
+  FROM pg_catalog.pg_class c
+  JOIN pg_catalog.pg_namespace n ON c.relnamespace=n.oid
+ WHERE c.relkind = 'i'
+SQL
+ ORDER BY n.nspname, c.relname
+SQL
+    return;
+}
+
+sub rebuild {
+
+    my ( $options, $opt_validate );
+
+    ( $dbh, $options, $opt_dryrun ) = @_;
+
+    $opt_validate     = $options->{Validate};
+    $opt_throttle_on  = $options->{ThrottleOn} || 10000000;
+    $opt_throttle_off = $options->{ThrottleOff} || 100000;
+
+    while ( my @idx = next_index() ) {
+        reindex( @idx, $opt_validate );
+        query '', $idx[0],
+            q{UPDATE reindex.worklist SET status='done' WHERE idx=$1};
+    }
+}
+
+=head1 AUTHOR
+
+Torsten Förtsch, C<< <binary at cpan.org> >>
+
+=head1 BUGS
```

Comment:
The license below is ok. But can we just delete all the stuff from here to there? Nobody of us looks there. And to refer to the same page for support does not make sense.